go-controller/pkg/ovn/egressip.go

```diff
@@ -1005,106 +1127,321 @@ func (e *EgressIPController) syncLocalNodeZonesCache() error {
+			}
+		}
+		return ops, nil
 	}
-	var allEIPServedPodIPs []net.IP
-	// we only care about local zone pods for the address-set since
-	// traffic from remote pods towards nodeIP won't even reach this zone
-	for eipName := range egressIPCache {
-		for _, podIPs := range egressIPCache[eipName].egressLocalPods {
-			for podIP := range podIPs {
-				allEIPServedPodIPs = append(allEIPServedPodIPs, net.ParseIP(podIP))
+
+	var ops []ovsdb.Operation
+	for eIPName, networkPodCache := range egressIPCache.egressIPNameToPods {
+		if egressIPCache.markCache[eIPName] == "" {
+			continue
+		}
+		for networkName, podCache := range networkPodCache {
+			for eIP, nodeName := range egressIPCache.egressIPIPToNodeCache {
+				if !egressIPCache.egressLocalNodesCache.Has(nodeName) {
+					continue
+				}
+				ni, err := util.NewNetInfo(&ovncnitypes.NetConf{
+					Topology: types.Layer3Topology,
+					NetConf: cnitypes.NetConf{
+						Name: networkName,
+					},
+				})
+				if err != nil {
+					return fmt.Errorf("failed to create new network %s: %v", networkName, err)
+				}
+				routerName := ni.GetNetworkScopedGWRouterName(nodeName)
+				isEIPIPv6 := utilnet.IsIPv6String(eIP)
+				for podKey, podIPs := range podCache.egressLocalPods {
+					ops, err = processPodFn(ops, eIPName, podKey, egressIPCache.markCache[eIPName], routerName, networkName, podIPs, isEIPIPv6)
+					if err != nil {
+						return fmt.Errorf("failed to add process local pod pod %s gateway router SNAT mark: %v", podKey, err)
+					}
+				}
+				for podKey, podIPs := range podCache.egressRemotePods {
+					ops, err = processPodFn(ops, eIPName, podKey, egressIPCache.markCache[eIPName], routerName, networkName, podIPs, isEIPIPv6)
+					if err != nil {
+						return fmt.Errorf("failed to add process remote pod pod %s gateway router SNAT mark: %v", podKey, err)
+					}
+				}
 			}
 		}
 	}
-	// we replace all IPs in the address-set based on eIP cache constructed from kapi
-	// note that setIPs is not thread-safe
-	if err = as.SetAddresses(util.StringSlice(allEIPServedPodIPs)); err != nil {
-		return fmt.Errorf("cannot reset egressPodIPs in address set %v: err: %v", EgressIPServedPodsAddrSetName, err)
+	_, err := libovsdbops.TransactAndCheck(e.nbClient, ops)
+	if err != nil {
+		return fmt.Errorf("error transacting ops %+v: %v", ops, err)
 	}
+
 	return nil
 }
 
 // syncPodAssignmentCache rebuilds the internal pod cache used by the egressIP feature.
-// We use the existing kapi and ovn-db information to populate oc.podAssignment cache for
+// We use the existing kapi and ovn-db information to populate oc.eIPC.podAssignment cache for
 // all the pods that are managed by egressIPs.
 // NOTE: This is done mostly to handle the corner case where one pod has more than one
 // egressIP object matching it, in which case we do the ovn setup only for one of the objects.
 // This corner case  of same pod matching more than one object will not work for IC deployments
```

Comment:
while you are here, should this say:

This corner case of the same pod matching more than one object will is not applicable in IC deployments

?

Reply:
it already says this - i dont understand

Reply:
I rephrased it and got rid of a space. The rephrase was just to say the case is not relevant to IC deployments rather than "something is broken/not work" in IC deployments. not a big deal though if you dont fix that in your PR.